go/extractor/util/registryproxy.go

```diff
@@ -0,0 +1,140 @@
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+const PROXY_HOST = "CODEQL_PROXY_HOST"
+const PROXY_PORT = "CODEQL_PROXY_PORT"
+const PROXY_CA_CERTIFICATE = "CODEQL_PROXY_CA_CERTIFICATE"
+const PROXY_URLS = "CODEQL_PROXY_URLS"
+const GOPROXY_SERVER = "goproxy_server"
+
+type RegistryConfig struct {
+	Type string `json:"type"`
+	URL  string `json:"url"`
+}
+
+// The address of the proxy including protocol and port (e.g. http://localhost:1234)
+var proxy_address string
+
+// The path to the temporary file that stores the proxy certificate, if any.
+var proxy_cert_file string
+
+// An array of registry configurations that are relevant to Go.
+// This excludes other registry configurations that may be available, but are not relevant to Go.
+var proxy_configs []RegistryConfig
+
+// Stores the environment variables that we wish to pass on to `go` commands.
+var proxy_vars []string = nil
+
+// Keeps track of whether we have inspected the proxy environment variables.
+// Needed since `proxy_vars` may be nil either way.
+var proxy_vars_checked bool = false
+
+// Tries to parse the given string value into an array of RegistryConfig values.
+func parseRegistryConfigs(str string) ([]RegistryConfig, error) {
+	var configs []RegistryConfig
+	if err := json.Unmarshal([]byte(str), &configs); err != nil {
+		return nil, err
+	}
+
+	return configs, nil
+}
+
+func getEnvVars() []string {
+	// If `proxy_vars` has been initialised, then we have already performed
+	// these checks and don't need to do so again. We assume that the environment
+	// variables are constant throughout the run of the autobuilder.
+	if proxy_vars != nil {
+		return proxy_vars
+	}
+
+	var result []string
+
+	if proxy_host, proxy_host_set := os.LookupEnv(PROXY_HOST); proxy_host_set {
+		if proxy_port, proxy_port_set := os.LookupEnv(PROXY_PORT); proxy_port_set {
+			proxy_address = fmt.Sprintf("http://%s:%s", proxy_host, proxy_port)
+			result = append(result, fmt.Sprintf("HTTP_PROXY=%s", proxy_address), fmt.Sprintf("HTTPS_PROXY=%s", proxy_address))
+
+			slog.Info("Found private registry proxy", slog.String("proxy_address", proxy_address))
+		}
+	}
+
+	if proxy_cert, proxy_cert_set := os.LookupEnv(PROXY_CA_CERTIFICATE); proxy_cert_set {
+		// Write the certificate to a temporary file
+		slog.Info("Found certificate")
+
+		f, err := os.CreateTemp("", "codeql-proxy.crt")
+		if err != nil {
+			slog.Error("Unable to create temporary file for the proxy certificate", slog.String("error", err.Error()))
```

Comment:
```suggestion
			slog.Error("Failed to create temporary file for the proxy certificate", slog.String("error", err.Error()))
```
(nitpick) This is slightly more consistent with the other error messages.